File: app/src/test/java/org/ole/planet/myplanet/ui/teams/leaderboard/TeamLeaderboardCalculatorTest.kt
package org.ole.planet.myplanet.ui.teams.leaderboard

import com.google.gson.JsonObject
import java.util.Calendar
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Test
import org.ole.planet.myplanet.model.RealmUser
import org.ole.planet.myplanet.repository.JoinedMemberData

class TeamLeaderboardCalculatorTest {

    private fun progress(max: Int, current: Int, completedAt: Long = 0L): JsonObject {
        val obj = JsonObject()
        obj.addProperty("max", max)
        obj.addProperty("current", current)
        obj.addProperty("completedAt", completedAt)
        return obj
    }

    private fun member(id: String, name: String): JoinedMemberData {
        val user = RealmUser().apply {
            this.id = id
            this.name = name
        }
        return JoinedMemberData(
            user = user,
            visitCount = 0,
            lastVisitDate = null,
            offlineVisits = "",
            profileLastVisit = "",
            isLeader = false
        )
    }

    @Test
    fun `ranks member with more completed courses first`() {
        val alice = member("alice", "Alice")
        val bob = member("bob", "Bob")
        val courseIds = listOf("course1", "course2")
        val progressByUser = mapOf(
            "alice" to mapOf<String?, JsonObject>(
                "course1" to progress(max = 5, current = 5),
                "course2" to progress(max = 5, current = 5)
            ),
            "bob" to mapOf<String?, JsonObject>(
                "course1" to progress(max = 5, current = 5),
                "course2" to progress(max = 5, current = 1)
            )
        )

        val result = TeamLeaderboardCalculator.build(listOf(bob, alice), courseIds, progressByUser)

        assertEquals(listOf("alice", "bob"), result.map { it.userId })
        assertEquals(2, result[0].coursesCompleted)
        assertEquals(1, result[1].coursesCompleted)
    }

    @Test
    fun `breaks ties on completed courses using surveys completed`() {
        val alice = member("alice", "Alice")
        val bob = member("bob", "Bob")
        val courseIds = listOf("course1")
        val progressByUser = mapOf(
            "alice" to mapOf<String?, JsonObject>("course1" to progress(max = 5, current = 5)),
            "bob" to mapOf<String?, JsonObject>("course1" to progress(max = 5, current = 5))
        )
        val surveyTimestampsByUser = mapOf(
            "alice" to listOf(1L, 2L, 3L),
            "bob" to listOf(1L)
        )

        val result = TeamLeaderboardCalculator.build(
            listOf(bob, alice),
            courseIds,
            progressByUser,
            surveyTimestampsByUser,
            surveysTotal = 3
        )

        assertEquals(listOf("alice", "bob"), result.map { it.userId })
        assertEquals(3, result[0].surveysCompleted)
    }

    @Test
    fun `This month period excludes courses completed before the period start`() {
        val alice = member("alice", "Alice")
        val courseIds = listOf("course1", "course2")
        val progressByUser = mapOf(
            "alice" to mapOf<String?, JsonObject>(
                "course1" to progress(max = 5, current = 5, completedAt = 2_000L),
                "course2" to progress(max = 5, current = 5, completedAt = 500L)
            )
        )

        val allTime = TeamLeaderboardCalculator.build(listOf(alice), courseIds, progressByUser, periodStart = null)
        val thisMonth = TeamLeaderboardCalculator.build(listOf(alice), courseIds, progressByUser, periodStart = 1_000L)

        assertEquals(2, allTime[0].coursesCompleted)
        assertEquals(1, thisMonth[0].coursesCompleted)
    }

    @Test
    fun `This month period excludes surveys completed before the period start`() {
        val alice = member("alice", "Alice")
        val surveyTimestampsByUser = mapOf("alice" to listOf(500L, 1_500L, 2_500L))

        val allTime = TeamLeaderboardCalculator.build(
            listOf(alice), emptyList(), emptyMap(), surveyTimestampsByUser, surveysTotal = 3, periodStart = null
        )
        val thisMonth = TeamLeaderboardCalculator.build(
            listOf(alice), emptyList(), emptyMap(), surveyTimestampsByUser, surveysTotal = 3, periodStart = 1_000L
        )

        assertEquals(3, allTime[0].surveysCompleted)
        assertEquals(2, thisMonth[0].surveysCompleted)
    }

    @Test
    fun `marks the entry matching currentUserId as the current user`() {
        val alice = member("alice", "Alice")
        val bob = member("bob", "Bob")

        val result = TeamLeaderboardCalculator.build(
            listOf(alice, bob), emptyList(), emptyMap(), currentUserId = "bob"
        )

        assertEquals(false, result.first { it.userId == "alice" }.isCurrentUser)
        assertEquals(true, result.first { it.userId == "bob" }.isCurrentUser)
    }

    @Test
    fun `member with no progress data still appears with zeroed totals`() {
        val alice = member("alice", "Alice")
        val courseIds = listOf("course1")

        val result = TeamLeaderboardCalculator.build(listOf(alice), courseIds, emptyMap())

        assertEquals(1, result.size)
        assertEquals(0, result[0].coursesCompleted)
        assertEquals(1, result[0].coursesTotal)
    }

    @Test
    fun `falls back to name when first and last name are blank`() {
        val guest = member("guest1", "guestname")

        val result = TeamLeaderboardCalculator.build(listOf(guest), emptyList(), emptyMap())

        assertEquals("guestname", result[0].displayName)
    }

    @Test
    fun `startOfCurrentMonth returns the first instant of the month containing now`() {
        val now = System.currentTimeMillis()

        val start = TeamLeaderboardCalculator.startOfCurrentMonth(now)

        assertTrue(start <= now)
        val calendar = Calendar.getInstance()
        calendar.timeInMillis = start
        assertEquals(1, calendar.get(Calendar.DAY_OF_MONTH))
        assertEquals(0, calendar.get(Calendar.HOUR_OF_DAY))
        assertEquals(0, calendar.get(Calendar.MINUTE))
        assertEquals(0, calendar.get(Calendar.SECOND))
    }
}

File: flutter/test/ui/teams/leaderboard/team_leaderboard_calculator_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:myplanet/ui/teams/leaderboard/team_leaderboard_calculator.dart';

import 'package:myplanet/repository/progress_repository.dart';

void main() {
  CourseProgressSummary summary(int current, int max) =>
      CourseProgressSummary(current: current, max: max);

  group('TeamLeaderboardCalculator', () {
    test('ranks by courses completed then surveys completed', () {
      final members = [
        const MemberWithProgress(
          userId: 'a',
          displayName: 'Alice',
          visitCount: 5,
        ),
        const MemberWithProgress(
          userId: 'b',
          displayName: 'Bob',
          visitCount: 3,
        ),
        const MemberWithProgress(
          userId: 'c',
          displayName: 'Carol',
          visitCount: 1,
        ),
      ];
      final courseIds = ['course-1', 'course-2'];

      final entries = TeamLeaderboardCalculator.build(
        members: members,
        courseIds: courseIds,
        progressForUser: (userId) => switch (userId) {
          'a' => {'course-1': summary(5, 5), 'course-2': summary(3, 5)},
          'b' => {'course-1': summary(5, 5), 'course-2': summary(5, 5)},
          _ => {},
        },
        surveyTimestampsForUser: (userId) => switch (userId) {
          'a' => [1000, 2000],
          'b' => [1000],
          _ => [],
        },
        surveysTotal: 3,
        currentUserId: 'a',
      );

      // Bob completes 2 courses → rank 1, Alice 1 → rank 2, Carol 0 → rank 3.
      expect(entries.map((e) => e.userId).toList(), ['b', 'a', 'c']);
      expect(entries[0].coursesCompleted, 2);
      expect(entries[1].coursesCompleted, 1);
      expect(entries[1].surveysCompleted, 2);
      expect(entries[1].isCurrentUser, isTrue);
    });

    test('thisMonth period filters survey timestamps', () {
      final members = [
        const MemberWithProgress(
          userId: 'a',
          displayName: 'Alice',
          visitCount: 0,
        ),
      ];
      final now = DateTime.utc(2026, 8, 15).millisecondsSinceEpoch;
      final periodStart = TeamLeaderboardCalculator.startOfCurrentMonth(now);

      final entries = TeamLeaderboardCalculator.build(
        members: members,
        courseIds: const [],
        progressForUser: (_) => {},
        surveyTimestampsForUser: (_) => [
          DateTime.utc(2026, 7, 1).millisecondsSinceEpoch,
          periodStart + 1,
        ],
        surveysTotal: 5,
        currentUserId: null,
        periodStart: periodStart,
      );

      // Only the August timestamp counts; the July one is filtered.
      expect(entries.first.surveysCompleted, 1);
    });

    test('startOfCurrentMonth is the first millisecond of the month', () {
      final now = DateTime.utc(2026, 8, 24, 14, 30).millisecondsSinceEpoch;
      final start = TeamLeaderboardCalculator.startOfCurrentMonth(now);
      final startDt = DateTime.fromMillisecondsSinceEpoch(start, isUtc: true);
      expect(startDt.year, 2026);
      expect(startDt.month, 8);
      expect(startDt.day, 1);
      expect(startDt.hour, 0);
    });

    test('empty members list returns empty', () {
      final entries = TeamLeaderboardCalculator.build(
        members: const [],
        courseIds: const [],
        progressForUser: (_) => {},
        surveyTimestampsForUser: (_) => const [],
        surveysTotal: 0,
        currentUserId: null,
      );
      expect(entries, isEmpty);
    });
  });
}
